Google.Api.Gax.Grpc/Rest/RestChannel.cs

```diff
@@ -102,14 +127,28 @@ private async Task<ReadHttpResponseMessage> SendAsync<TRequest>(RestMethod restM
             }
         }
 
-        private async Task AddAuthHeadersAsync(HttpRequestMessage request, RestMethod method)
+        private async Task AddAuthHeadersAsync(HttpRequestMessage request, RestMethod restMethod, CallOptions options, CancellationToken deadlineToken)
         {
             Uri hostUri = request.RequestUri.IsAbsoluteUri ? request.RequestUri : _httpClient.BaseAddress;
             string schemeAndAuthority = hostUri.GetLeftPart(UriPartial.Authority);
 
             var metadata = new Metadata();
-            var context = new AuthInterceptorContext(schemeAndAuthority, method.FullName);
-            await _channelAuthInterceptor(context, metadata).ConfigureAwait(false);
+            var context = new AuthInterceptorContext(schemeAndAuthority, restMethod.FullName);
+
+            var deadlineTask = Task.Delay(-1, deadlineToken);
+            var clientCancellationTask = Task.Delay(-1, options.CancellationToken);
+            var channelTask = _channelAuthInterceptor(context, metadata);
+            var resultTask = await Task.WhenAny(channelTask, deadlineTask, clientCancellationTask).ConfigureAwait(false);
+            if (deadlineTask == resultTask)
+            {
+                throw new RpcException(new Status(StatusCode.DeadlineExceeded, $"Timeout was reached when waiting for auth headers to be added for a method `{restMethod.FullName}`."));
+            }
+
+            if (clientCancellationTask == resultTask)
```

Comment:
channelTask may have failed as well, in which case we are not failing, we are just not adding any headers, I guess we should buble the channelTask failure if there was any?

I wonder, why don't you pass the linked token as a parameter here and then do something like:

if (resultTask.IsFaulted)
{
    throw resultTask.Exception; 
}

And then you catch that the same way you catch the one in the calling _httpClient.SendAsync(...). If you want to keep the different messages you can set a flag after you have added the headers to diferentiate which operation threw.

Reply:
I've implemented a version of this.